### PR TITLE
Multiple machines per token

### PR DESCRIPTION
**Pull request by zalun**
this is an ongoing process. Starts with back-end

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -7,7 +8,13 @@ gulp.task('clean', function(done) {
   done();
 });
 
-gulp.task('default', ['clean'], function() {
+gulp.task('lint', function() {
+  return gulp.src(['./*.js', './static/*.js'])
```
Comment (marco-c) [inline gulpfile.js]: You could use *_/_.js to select all JS files.

File: index.js [hunk 2]
```diff
@@ -39,20 +42,65 @@ if (!fs.existsSync('./dist')) {
...
-  crypto.randomBytes(32, function(ex, buf) {
-    var token = buf.toString('hex');
+// load current data for
+app.get('/', function(req, res) {
+  res.render('index', { title: 'Hey', message: 'Hello there!'});
+});
 
-    // save token in database
-    client.hmset(token, {
+// adds a new machine and adds to a token set
```
Comment (marco-c) [inline index.js]: "adds a new machine to a token set"?

File: index.js [hunk 3]
```diff
@@ -39,20 +42,65 @@ if (!fs.existsSync('./dist')) {
...
+// creates a new token set if needed
+app.post('/register', function(req, res) {
+  // add/update machine in database
+  var machineId = req.body.machineId;
+  new Promise(function(allfine, reject) {
```
Comment (marco-c) [inline index.js]: Nit: can you use the canonical "resolve" instead of "allfine"?

File: test/test.js [hunk 4]
```diff
@@ -42,6 +44,54 @@ describe('mercurius', function() {
...
+  it('replies with 404 when trying to unregister a non registered machine', function(done) {
+    request(mercurius.app)
+      .post('/unregisterMachine')
+      .send({
+        token: tokenToUnregister,
+        machineId: 'non-existing-machine',
+      })
+      .expect(404, done);
+  });
+
```
Comment (marco-c) [inline test/test.js:126]: We should also test that notify sends the notification to all machines in a token set.

File: static/index.js [hunk 5]
```diff
@@ -27,48 +45,81 @@ function register() {
...
-
-      localforage.clear();
+      domToken.textContent = '';
+      showSection('registrationForm');
+      localforage.removeItem('token');
     });
   });
 };
 
+function makeId(length) {
```
Comment (marco-c) [inline static/index.js:100]: You could use: https://developer.mozilla.org/en-US/docs/Web/API/RandomSource/getRandomValues#Example

File: index.js [hunk 6]
```diff
@@ -39,20 +42,65 @@ if (!fs.existsSync('./dist')) {
...
+      return allfine();
+    }
+    client.exists(req.body.token, function(err, result) {
+      if (!result || err) {
+        return reject();
+      }
+      allfine();
+    });
+  }).then(function() {
+    client.hmset(machineId, {
```
Comment (marco-c) [inline index.js:66]: Does this mean that two users can't have the same machine ID?

File: static/service-worker.js [hunk 7]
```diff
@@ -13,24 +13,33 @@ self.addEventListener('push', function(event) {
...
+    .then(function(token) {
       if (!token) {
         return;
       }
 
-      return self.registration.pushManager.subscribe({ userVisibleOnly: true }).then(function(subscription) {
-        var key = subscription.getKey ? subscription.getKey('p256dh') : '';
+      localforage.getItem('machineId')
+      .then(function(machineId) {
+        if (!token) {
```
Comment (marco-c) [inline static/service-worker.js]: No need to check again here.

File: index.js (hunk 3 above)
Comment (zalun) [inline index.js]: yes - I had a promise inside promise before refactoring

File: static/index.js [hunk 8]
```diff
@@ -27,48 +45,81 @@ function register() {
...
+          name: domMachineName.value
         }),
       }).then(function(response) {
         response.text().then(function(token) {
-          localforage.setItem('token', token);
-          document.getElementById('token').textContent = token;
-          document.getElementById('registrationForm').style.display = 'none';
-          document.getElementById('unregistrationForm').style.display = 'block';
+          if (response.ok) {
+              localforage.setItem('token', token);
```
Comment (marco-c) [inline static/index.js]: Nit: fix indentation

File: index.js [hunk 9]
```diff
@@ -60,48 +108,138 @@ app.post('/unregister', function(req, res) {
...
-        "endpoint": req.body.endpoint,
-        "key": req.body.key
-      }, function() {
-        res.sendStatus(200);
+    client.sismember(token, machineId, function(err, ismember) {
+      if (!ismember) {
+        res.sendStatus(404);
+        return;
+      }
+      client.exists(machineId, function(err, exists) {
```
Comment (marco-c) [inline index.js]: Is this call needed after `sismember`?

File: index.js [hunk 10]
```diff
@@ -60,48 +108,138 @@ app.post('/unregister', function(req, res) {
...
       return;
     }
-    client.del(token, function(err) {
-      res.sendStatus(200);
+    client.smembers(token, function(err, machines) {
+      for (var index = 0; index < machines.length; index++) {
+        client.del(machines[index]);
+      }
+      client.del(token, function(err) {
+        res.sendStatus(200);
```
Comment (marco-c) [inline index.js]: We should finish after removing both machines and token.
Maybe if we wrapped the redis API in promises all the code in index.js would get cleaner, we can do that in a followup.

File: test/test.js (hunk 4 above)
Comment (zalun) [inline test/test.js:126]: Code to send notification should also be slightly different - it should send 200 only if all requests returned 200 (this is done now)

Comment (marco-c) [inline test/test.js:126]: There are probably other things that we should test, you can check the code coverage info to see what isn't being tested.

File: index.js [hunk 11]
```diff
@@ -60,48 +108,138 @@ app.post('/unregister', function(req, res) {
...
   });
 });
 
-app.post('/notify', function(req, res) {
+// used only if metadata updated
+// this happens on request to update meta data from front-end site
+app.post('/updateMeta', function(req, res) {
   var token = req.body.token;
+  var machineId = req.body.machineId;
   client.exists(token, function(err, exists) {
```
Comment (marco-c) [inline index.js:189]: This call should be unneeded, because the following `sismember` will fail if `token` doesn't exist.

Comment (marco-c) [inline index.js:189]: Let's also add a test of `updateMeta` when the token doesn't exist to make sure it works (and also when the machineId doesn't exist).

File: test/test.js (hunk 4 above)
Comment (marco-c) [inline test/test.js:126]: (Once you run `npm test`, open the file `coverage/lcov-report/mercurius/index.js.html`)

File: index.js (hunk 9 above)
Comment (zalun) [inline index.js]: yes - it means: some machineId is inside token set, but it doesn't mean that such hash exists in database

Comment (marco-c) [inline index.js]: ok!

File: gulpfile.js (hunk 1 above)
Comment (marco-c) [inline gulpfile.js]: The previous comment is cut off because of markdown I guess, to actually see it click `edit`.

File: index.js [hunk 12]
```diff
@@ -39,20 +42,65 @@ if (!fs.existsSync('./dist')) {
...
+      return resolve();
+    }
+    client.exists(req.body.token, function(err, result) {
+      if (!result || err) {
+        return reject();
+      }
+      resolve();
+    });
+  }).then(function() {
+    client.hmset(machineId, {
```
Comment (marco-c) [inline index.js:66]: In theory we should create machineId and token in a single transaction (using `multi`). Let's do it in a follow-up.

**Pull request merged**